src/dune/simple_rules.ml

```diff
@@ -57,6 +67,29 @@ let user_rule sctx ?extra_bindings ~dir ~expander (rule : Rule.t) =
     in
     let bindings = dep_bindings ~extra_bindings rule.deps in
     let expander = Expander.add_bindings expander ~bindings in
+    let action =
+      SC.Deps.interpret_named sctx ~expander rule.deps
+      |> SC.Action.run sctx (snd rule.action) ~loc:(fst rule.action) ~expander
+           ~dep_kind:Required ~targets ~targets_dir:dir
+    in
+    let action =
+      match rule.alias with
+      | None -> action
+      | Some alias ->
+        let stamp =
+          let action = Some (snd rule.action) in
+          stamp ~deps:rule.deps ~extra_bindings ~action
+          |> Digest.generic
+          |> Digest.to_string
+        in
+        let stamp_target = Path.Build.relative dir stamp in
```

Comment:
I expect a stamp target matching random globs to be pretty annoying. We should probably do something about that.

One (somewhat hacky) approach is, as I said before, to move the rule action to the alias dir if there's no targets. (I think alias actions are translated to rules internally in builds_system.ml anyway so it's perhaps not as much of a hack as it seems at first)

Another thing we could do is to allow targets in multiple directories as long as the set of directories is known in advance, but that seems trickier.

Reply:
> One (somewhat hacky) approach is, as I said before, to move the rule action to the alias dir if there's no targets. (I think alias actions are translated to rules internally in builds_system.ml anyway so it's perhaps not as much of a hack as it seems at first)

But what about rules with targets? We'd be left with the same problem in this case.

> Another thing we could do is to allow targets in multiple directories as long as the set of directories is known in advance, but that seems trickier.

This is highly desirable. It's required to support  other kinds of rules such as coq's native compute rules.

However, I have a simpler solution for this problem. What I think we need is a "private" notion of targets. Such private targets:

* Cannot be picked up by user written globs
* Cannot be used as explicit dependencies for user written rules.

Such targets would solve other similar problems:
* Users relying on implementation specific paths like `.objs` and then dune changes end up breaking them.
* They make it easy for us to exclude undesired stuff from `@all`.

Reply:
> But what about rules with targets?

For rules with targets we don't really need a stamp file because we can depend on any of those targets.

> This is highly desirable.

Yes indeed!

> "private" notion of targets

That idea only helps if the rules are sandboxed. If they are not sandboxed, they will see the entire directory and dune has no power over that. Imagine somebody writing a rule "cat *" and declaring that it depends on glob "*".

A safe solution is to make it an error when a glob for a non-sandboxed rule happens to include a private dep, but that's obviously not great.